README
SmokeBox
========

A POE-based framework for automated CPAN Testing.

Supports multiple backends, including:

	- CPAN::YACSmoke
	- CPANPLUS::YACSmoke
	- CPAN::Reporter

Installation:
=============

The minimum version of perl supported is 5.6. This does not preclude CPAN testing of 
perls less than this, as the testing framework and the testing targets are two distinct
environments.

The included test-suite requires the following modules:

	Test::More 0.47
	File::Spec 0

The following modules are required:

	POE 1.0001
	Module::Pluggable 0
	Params::Check 0
	Object::Accessor 0

The recommended method of installation as always is to use either CPAN or CPANPLUS 
so that the dependencies are automagically dealt with for you and you have the option
of sending test reports during installation >:)

Status:
=======

Currently a developer release. Feel free to have a look, try it out, etc.

Any bug reports, feature requests, send an email to bug-POE-Component-SmokeBox@rt.cpan.org


Changes
==============================
2011-05-19 10:15:58 +0100 0.48
==============================

  commit 45ce9b17626a920a6e849f927424fa06edccbc88
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Thu May 19 10:15:58 2011 +0100
  
    Bump version and correct trailing whitespace.

  commit c1d8d64c26521111644def1eab74d1c1447b9558
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Thu May 19 10:12:20 2011 +0100
  
    Add a wheel reaper process to clean up the wheel if it is left open
    for some reason. 

==============================
2011-04-20 10:06:00 +0100 0.46
==============================

  commit f6ac2eb489c151b29e40ff2ed42c7133d3112724
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Wed Apr 20 10:06:00 2011 +0100
  
    Bump version to 0.46

  commit a54f17b139ec187d5236b49c1a746bb82b0b7bbf
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Wed Apr 20 09:59:10 2011 +0100
  
    Use hex digests instead of binary ones, suppress debug lines > 5000
    and only check for perl warnings if the line < 5000 

==============================
2011-02-15 21:49:54 +0000 0.44
==============================

  commit 83287bf3428d6d48732d044959b0778201ce528e
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Tue Feb 15 21:49:54 2011 +0000
  
    Bump version to 0.44

  commit b2ead74e698c57990e3fa2288c65b8c80d86c13b
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Tue Feb 15 21:48:48 2011 +0000
  
    Tidied up the digests tie code 

=================================
2011-02-15 12:45:26 +0000 0.43_01
=================================

  commit 38edc7a62be6ef1ae1b7c10f9210a1879832bcb5
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Tue Feb 15 12:45:26 2011 +0000
  
    Bump to dev release 0.43_01

  commit 56a0afb0fb95e45b5271b3a3a2f9f44fbab6bd4e
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Tue Feb 15 12:43:36 2011 +0000
  
    Store the SHA256 digests in a SDBM_File backed hash to reduce memory
    footprint 

==============================
2011-02-06 23:00:55 +0000 0.42
==============================

  commit 9d8eae91fd34f31c4b06eb9ef343f3aec7811bb6
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Sun Feb 6 23:00:55 2011 +0000
  
    Add .gitignore

  commit ff4d5c448520094ac351cf876952cdd996596ffc
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Sun Feb 6 22:53:50 2011 +0000
  
    Bump to version 0.42

  commit 8de3fe614d431cbe8752347ec762485699a342ab
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Sun Feb 6 22:45:30 2011 +0000
  
    Change to Digest::SHA instead of Digest::MD5 for our looping output
    tracking 

==============================
2011-01-20 12:18:31 +0000 0.40
==============================

  commit 08d2121e51b378f6d9f90d69a1ef7f8f278e85a2
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Thu Jan 20 12:18:31 2011 +0000
  
    Remove the warning message from Makefile.PL

  commit 3c7a479756becfa716b73f56cae59b81b3fbbf3d
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Thu Jan 20 12:16:48 2011 +0000
  
    Bump version to 0.40

  commit 472a425c75859708b7f5b76cc7fffa8ee4104370
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Thu Jan 20 10:21:30 2011 +0000
  
    Remove Changes file and config for it to be automatically generated

  commit 21fb5c3ea463ed8c37838fc28630c694959a8c20
  Author: Apocalypse <perl@0ne.us>
  Date:   Thu Jan 20 09:51:28 2011 +0800
  
    Add pause_queue_now() to pause between smokers and more testcases for
    the pause logic 

==============================
2010-04-27 21:54:00 +0100 0.38
==============================

  commit 1441441f79e94a54c970ad6e8030d4b2c8b45829
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date:   Tue Apr 27 21:54:00 2010 +0100
  
    Bump version to 0.38

  commit 6173f8e97ac9f915a1d4c525bb1f208073d276de
  Author: Apocalypse <perl@0ne.us>
  Date:   Tue Apr 27 10:07:04 2010 -0700
  
    add name to SmokeBox::Smoker for easier identification of smokers

  commit bfac61471cd2b7d465dafe0b0b2629dff1d8df76
  Author: Apocalypse <perl@0ne.us>
  Date:   Tue Apr 20 11:41:04 2010 -0700
  
    add Win32API::File code to disable popups during smoking

  commit 7f951cb853fe162cd8807b7d2a377d608cb261f1
  Author: Apocalypse <perl@0ne.us>
  Date:   Mon Apr 19 02:34:54 2010 -0700
  
    add check_warnings option to SmokeBox::Job->new to control usage of
    String::Perl::Warnings

  commit c1ee77ecf00f8712d8f629930307c96385e2e4fe
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date:   Fri Apr 16 20:57:40 2010 +0100
  
    Explicitly use the \n InputLiteral in Backend's PWR

  commit d7e967927b0682c54a400685789130dd71c6760e
  Author: Apocalypse <perl@0ne.us>
  Date:   Sat Apr 10 21:48:05 2010 -0700
  
    d0h, POE::Session->callback returns POE::Session::AnonEvent, not a
    pure CODE ref

  commit fa00ab4d4c9a78e9da9cb158648794bff402d2da
  Author: Apocalypse <perl@0ne.us>
  Date:   Sat Apr 10 18:44:14 2010 -0700
  
    reverse the wrong logic in IO::Pty detection

  commit 1fa1dd67a3dc06a4a103c01042495ffd30d0a7be
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date:   Thu Apr 8 22:19:46 2010 +0100
  
    Makefile.PL jiggery for MSWin32

  commit 96357ee17162902515dbede7fbe2335d131901a0
  Author: Apocalypse <perl@0ne.us>
  Date:   Wed Apr 7 22:37:02 2010 -0700
  
    Added callback mode to the SmokeBox::Smoker object for extensibility 

==============================
2010-03-22 22:57:56 +0000 0.36
==============================

  commit 749ab5f3b9ce0adf34be784c075724b304b87891
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date:   Mon Mar 22 22:57:56 2010 +0000
  
    This be 0.36

  commit 863866e177f85277f77498d5383716643ec4199f
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date:   Mon Mar 22 22:30:11 2010 +0000
  

examples/mini-smokebox.pl
  use strict;
  use warnings;
  use POE;
  use POE::Component::SmokeBox;
  use POE::Component::SmokeBox::Smoker;
  use POE::Component::SmokeBox::Job;
  use POE::Component::SmokeBox::Recent;
  use Getopt::Long;
  
  $|=1;
  
  my ($perl, $jobs, $recenturl);
  
  GetOptions( 'perl=s' => \$perl, 'jobs=s' => \$jobs, 'recenturl' => \$recenturl );
  
  my @pending;
  if ( $jobs ) {
    open my $fh, "<$jobs" or die "$jobs: $!\n";
    while (<$fh>) {
          chomp;
          push @pending, $_;
    }
    close($fh);
  }
  
  my $smokebox = POE::Component::SmokeBox->spawn();
  
  POE::Session->create(
        package_states => [
           'main' => [ qw(_start _stop _results _recent) ],
        ],
        heap => { perl => $perl, pending => \@pending },
  );
  
  $poe_kernel->run();
  exit 0;
  
  sub _start {
    my ($kernel,$heap) = @_[KERNEL,HEAP];

    my $smoker = POE::Component::SmokeBox::Smoker->new(
	perl => $perl,
    );

    $smokebox->add_smoker( $smoker );

    if ( @{ $heap->{pending} } ) {
      $smokebox->submit( event => '_results', 
			 job => POE::Component::SmokeBox::Job->new( command => 'smoke', module => $_ ) )
        			for @{ $heap->{pending} };
    }
    else {
      POE::Component::SmokeBox::Recent->recent( 
          url => $recenturl || 'http://www.cpan.org/',
          event => '_recent',
      );
    }
    undef;
  }
  
  sub _stop {
    $smokebox->shutdown();
    undef;
  }
  
  sub _results {
    my $result = $_[ARG0];
    print $_, "\n" for map { @{ $_->{log} } } $results->{result}->results();
    undef;
  }

  sub _recent {
    my ($kernel,$heap,$job) = @_[KERNEL,HEAP,ARG0];
    die $job->{error}, "\n" if $job->{error};
      $smokebox->submit( event => '_results', 
			 job => POE::Component::SmokeBox::Job->new( command => 'smoke', module => $_ ) )
        			for @{ $job->{recent} };
    undef;
  }


inc/Module/Install/With.pm
#line 1
package Module::Install::With;

# See POD at end for docs

use strict;
use Module::Install::Base;

use vars qw{$VERSION $ISCORE @ISA};
BEGIN {
	$VERSION = '0.84';
	$ISCORE  = 1;
	@ISA     = qw{Module::Install::Base};
}





#####################################################################
# Installer Target

# Are we targeting ExtUtils::MakeMaker (running as Makefile.PL)
sub eumm {
	!! ($0 =~ /Makefile.PL$/i);
}

# You should not be using this, but we'll keep the hook anyways
sub mb {
	!! ($0 =~ /Build.PL$/i);
}





#####################################################################
# Testing and Configuration Contexts

#line 49

sub interactive {
	# Treat things interactively ONLY based on input
	!! (-t STDIN and ! automated_testing());
}

#line 67

sub automated_testing {
	!! $ENV{AUTOMATED_TESTING};
}

#line 86

sub release_testing {
	!! $ENV{RELEASE_TESTING};
}

sub author_context {
	!! $Module::Install::AUTHOR;
}





#####################################################################
# Operating System Convenience

#line 114

sub win32 {
	!! ($^O eq 'MSWin32');
}

#line 131

sub winlike {
	!! ($^O eq 'MSWin32' or $^O eq 'cygwin');
}

1;

#line 159


lib/POE/Component/SmokeBox.pm
package POE::Component::SmokeBox;

use strict;
use warnings;
use POE qw(Component::SmokeBox::Backend Component::SmokeBox::JobQueue);
use POE::Component::SmokeBox::Smoker; 
use POE::Component::SmokeBox::Job;
use POE::Component::SmokeBox::Result;
use vars qw($VERSION);

$VERSION = '0.16';

sub spawn {
  my $package = shift;
  my %params = @_;
  $params{lc $_} = delete $params{$_} for keys %params;
  my $options = delete $params{'options'};
  my $self = bless \%params, $package;
  $self->{session_id} = POE::Session->create(
	object_states => [
	   $self => { 
		shutdown   => '_shutdown',
		add_smoker => '_add_smoker',
		del_smoker => '_del_smoker',
		submit     => '_submit',
	   },
	   $self => [qw(_start)],
	],
	heap => $self,
	( ref($options) eq 'HASH' ? ( options => $options ) : () ),
  )->ID();
  return $self;
}

sub session_id {
  return $_[0]->{session_id};
}

sub multiplicity {
  return $_[0]->{multiplicity};
}

sub queues {
  return map { $_->{queue} } @{ $_[0]->{queues} };
}

sub shutdown {
  my $self = shift;
  $poe_kernel->call( $self->session_id() => 'shutdown' => @_ );
}

sub _start {
  my ($kernel,$self) = @_[KERNEL,OBJECT];
  $self->{session_id} = $_[SESSION]->ID();
  if ( $self->{alias} ) {
    $kernel->alias_set( $self->{alias} );
  }
  else {
    $kernel->refcount_increment( $self->{session_id} => __PACKAGE__ );
  }
  $self->{queues} = [ ];
  my $smokers = delete $self->{smokers};
  return unless $smokers and ref $smokers eq 'ARRAY' and scalar @{ $smokers };
  $self->add_smoker( $_ ) for @{ $smokers };
  return;
}

sub _shutdown {
  my ($kernel,$self) = @_[KERNEL,OBJECT];
  if ( $self->{alias} ) {
	$kernel->alias_remove($_) for $kernel->alias_list();
  } 
  else {
	$kernel->refcount_decrement( $self->{session_id} => __PACKAGE__ );
  }
  $_->{queue}->shutdown() for @{ $self->{queues} };
  return;
}

sub add_smoker {
  my $self = shift;
  $poe_kernel->call( $self->{session_id}, 'add_smoker', @_ );
}

sub del_smoker {
  my $self = shift;
  $poe_kernel->call( $self->{session_id}, 'del_smoker', @_ );
}

sub _add_smoker {
  my ($kernel,$self,$state,$sender,$smoker) = @_[KERNEL,OBJECT,STATE,SENDER,ARG0];
  unless ( $smoker and $smoker->isa('POE::Component::SmokeBox::Smoker') ) {
     warn "ARG0 must be a 'POE::Component::SmokeBox::Smoker' object\n";
     return;
  }
  # If no jobqueues start a job queue.
  # If multiplicity start a job queue for each smoker object.
  if ( $self->{multiplicity} or scalar @{ $self->{queues} } == 0 ) {
    my $queue = { };
    $queue->{queue} = POE::Component::SmokeBox::JobQueue->spawn();
    push @{ $queue->{smokers} }, $smoker;
    push @{ $self->{queues} }, $queue;
    return;
  }
  # Otherwise we just add the smoker to our existing queue
  push @{ $self->{queues}->[0]->{smokers} }, $smoker;
  return;
}

sub _del_smoker {
  my ($kernel,$self,$state,$sender,$smoker) = @_[KERNEL,OBJECT,STATE,SENDER,ARG0];
  unless ( $smoker and $smoker->isa('POE::Component::SmokeBox::Smoker') ) {
     warn "ARG0 must be a 'POE::Component::SmokeBox::Smoker' object\n";
     return;
  }
  my $x = 0;
  foreach my $queue ( @{ $self->{queues} } ) {
     my $i = 0;
     for ( @{ $queue->{smokers} } ) {
        splice(@{ $queue->{smokers} }, $i, 1) if $_ == $smoker;
        ++$i;
     }
     unless ( scalar @{ $queue->{smokers} } ) {
        splice(@{ $self->{queues} }, $x, 1);
	$queue->{queue}->shutdown();
     }
     ++$x;
  }
  return;
}

sub submit {
  my $self = shift;
  $poe_kernel->call( $self->{session_id}, 'submit', @_ );
}

sub _submit {
  my ($kernel,$self,$state,$sender) = @_[KERNEL,OBJECT,STATE,SENDER];
  return if $self->{_shutdown};
  my $args;
  if ( ref( $_[ARG0] ) eq 'HASH' ) {
     $args = { %{ $_[ARG0] } };
  } 
  else {
     $args = { @_[ARG0..$#_] };
  }

  $args->{lc $_} = delete $args->{$_} for grep { $_ !~ /^_/ } keys %{ $args };

  unless ( $args->{event} ) {
     warn "No 'event' specified for $state\n";
     return;
  }

  unless ( $args->{job} and $args->{job}->isa('POE::Component::SmokeBox::Job') ) {
     warn "No 'job' specified for $state or it was not a valid 'POE::Component::SmokeBox::Job' object\n";
     return;
  }

  if ( $args->{session} and my $ref = $kernel->alias_resolve( $args->{session} ) ) {
     $args->{session} = $ref->ID();
  }
  else {
     $args->{session} = $sender->ID();
  }
  
  warn "No smokers have been defined yet!!!!!\n" unless scalar @{ $self->{queues} };
  
  foreach my $q ( @{ $self->{queues} } ) {
     $args->{smokers} = [ @{ $q->{smokers} } ];
     $q->{queue}->submit( $args );
  }

  return;
}

"We've Got a Fuzzbox and We're Gonna Use It";

__END__

=head1 NAME

POE::Component::SmokeBox - POE enabled CPAN smoke testing with added value.

=head1 SYNOPSIS

  # A simple smoker that takes modules to smoke from @ARGV

  use strict;
  use warnings;
  use POE;
  use POE::Component::SmokeBox;
  use POE::Component::SmokeBox::Smoker;
  use POE::Component::SmokeBox::Job;
  use Getopt::Long;
  
  $|=1;
  
  my $perl;

  GetOptions( 'perl=s' => \$perl, );

  die "No 'perl' specified\n" unless $perl;

  die "No modules specified to smoke\n" unless scalar @ARGV;
  
  my $smokebox = POE::Component::SmokeBox->spawn();
  
  POE::Session->create(
        package_states => [
           'main' => [ qw(_start _stop _results) ],
        ],
        heap => { perl => $perl, pending => [ @ARGV ] },
  );
  
  $poe_kernel->run();
  exit 0;
  
  sub _start {
    my ($kernel,$heap) = @_[KERNEL,HEAP];

    my $smoker = POE::Component::SmokeBox::Smoker->new( perl => $perl, );

    $smokebox->add_smoker( $smoker );

    $smokebox->submit( event => '_results', 
		 job => POE::Component::SmokeBox::Job->new( command => 'smoke', module => $_ ) )
       			for @{ $heap->{pending} };
    undef;
  }
  
  sub _stop {
    $smokebox->shutdown();
    undef;
  }
  
  sub _results {
    my $results = $_[ARG0];
    print $_, "\n" for map { @{ $_->{log} } } $results->{result}->results();
    undef;
  }

=head1 DESCRIPTION

POE::Component::SmokeBox is a flexible CPAN Smoke testing framework which provides an
extensible method for testing CPAN distributions against various different smoker backends.

A smoker backend is defined using a L<POE::Component::SmokeBox::Smoker> object and is basically
the path to a C<perl> executable that is configured for CPAN Testing and its associated environment settings.

The C<perl> executable must be configured appropriately to support CPAN testing with any of the currently 
supported backends, L<CPANPLUS::YACSmoke>, L<CPAN::YACSmoke> or L<CPAN::Reporter>. Additional backends may be
supported by inheriting and extending the backend base class L<POE::Component::SmokeBox::Backend::Base>.

By default, the component will test submitted jobs against each configured smoker in turn. Setting C<multiplicity>
to true will enable each job to be run against configured smokers in parallel.

=head1 CONSTRUCTOR

=over

=item C<spawn>

Creates a new session and returns an object. Takes a number of parameters:

  'alias', set an alias that you can use to address the component later;
  'options', a hashref of POE session options;
  'multiplicity', set to a true value to enable multiplicity, default is false;
  'smokers', an arrayref of POE::Component::SmokeBox::Smoker objects;
  
=back

=head1 METHODS

=over

=item C<session_id>

Returns the L<POE::Session> ID of the smokebox component.

=item C<multiplicity>

Returns true or false depending on whether multiplicity is enabled or not.

=item C<queues>

Returns a list of L<POE::Component::SmokeBox::JobQueue> objects that are currently active in the smokebox.

=item C<add_smoker>

Takes one mandatory argument, a L<POE::Component::SmokeBox::Smoker> object to add to the smokebox.

=item C<del_smoker>

Takes one mandatory argument, a L<POE::Component::SmokeBox::Smoker> object to remove from the smokebox.

=item C<submit>

Submits a job to the smokebox. Takes a number of parameters.

  'event', the event name where results should be sent, mandatory;
  'job', a POE::Component::SmokeBox::Job object to submit, mandatory;
  'session', optionally specify a different session to send the result event to;

=item C<shutdown>

Terminates the smokebox component. 

=back

=head1 INPUT EVENTS

=over

=item C<add_smoker>

Takes one mandatory argument, a L<POE::Component::SmokeBox::Smoker> object to add to the smokebox.

=item C<del_smoker>

Takes one mandatory argument, a L<POE::Component::SmokeBox::Smoker> object to remove from the smokebox.

=item C<submit>

Submits a job to the smokebox. Takes a number of parameters.

  'event', the event name where results should be sent, mandatory;
  'job', a POE::Component::SmokeBox::Job object to submit, mandatory;
  'session', optionally specify a different session to send the result event to;

=item C<shutdown>

Terminates the smokebox component. 

=back

=head1 OUTPUT EVENTS

An event will be sent on process completion with a hashref as C<ARG0>:

  'job', the POE::Component::SmokeBox::Job object of the job;
  'result', a POE::Component::SmokeBox::Result object containing the results;
  'submitted', the epoch time in seconds when the job was submitted;
  'event', the event that will be sent with the results;
  'session', the session ID the above event will be sent to;

The results will be same as returned by L<POE::Component::SmokeBox::Backend>. They may be obtained by querying the
L<POE::Component::SmokeBox::Result> object:

  $_[ARG0]->{result}->results() # produces a list

Each result is a hashref:

  'log', an arrayref of STDOUT and STDERR produced by the job;
  'PID', the process ID of the POE::Wheel::Run;
  'status', the $? of the process;
  'start_time', the time in epoch seconds when the job started running;
  'end_time', the time in epoch seconds when the job finished;
  'idle_kill', only present if the job was killed because of excessive idle;
  'excess_kill', only present if the job was killed due to excessive runtime;
  'term_kill', only present if the job was killed due to a poco shutdown event;

=head1 AUTHOR

Chris C<BinGOs> Williams <chris@bingosnet.co.uk>

=head1 LICENSE

Copyright (C) Chris Williams

This module may be used, modified, and distributed under the same terms as Perl itself. Please see the license that came with your Perl distribution for details.

=head1 SEE ALSO

L<POE::Component::SmokeBox::Smoker>

L<POE::Component::SmokeBox::Job>

L<POE::Component::SmokeBox::JobQueue>

L<POE::Component::SmokeBox::Backend>

L<POE::Component::SmokeBox::Result>

=cut


lib/POE/Component/SmokeBox/Job.pm
package POE::Component::SmokeBox::Job;

use strict;
use warnings;
use Params::Check qw(check);
use base qw(Object::Accessor);
use vars qw($VERSION $VERBOSE);

$VERSION = '0.01';

sub new {
  my $package = shift;

  my $tmpl = {
	idle    => { allow => qr/^\d+$/, default => 600, },
	timeout => { allow => qr/^\d+$/, default => 3600, },
	type    => { defined => 1, default => 'CPANPLUS::YACSmoke', },
	command => { allow => [ qw(check index smoke) ], default => 'check', },
	module  => { defined => 1 },
	smokers => { defined => 1, allow => [
		sub { 
			return 1 if ref $_[0] eq 'ARRAY'
				and scalar @{ $_[0] }
				and ( grep { $_->isa('POE::Component::SmokeBox::Smoker') } @{ $_[0] } ) == @{ $_[0] };
		}, 
	], },
  };

  my $args = check( $tmpl, { @_ }, 1 ) or return;
  if ( $args->{command} eq 'smoke' and !$args->{module} ) {
     warn "${package}::new expects 'module' to be set when command is 'smoke'";
     return;
  }
  my $self = bless { }, $package;
  my $accessor_map = {
	idle    => qr/^\d+$/,
	timeout => qr/^\d+$/,
	type    => sub { defined $_[0]; },
	command => [ qw(check index smoke) ],
	module  => sub { defined $_[0]; },
	smokers => 
		sub { 
			return 1 if ref $_[0] eq 'ARRAY'
				and scalar @{ $_[0] }
				and ( grep { $_->isa('POE::Component::SmokeBox::Smoker') } @{ $_[0] } ) == @{ $_[0] };
		}, 
	id	=> sub { defined $_[0]; },
  };
  $self->mk_accessors( $accessor_map );
  $self->$_( $args->{$_} ) for keys %{ $args };
  return $self;
}

sub dump_data {
  my $self = shift;
  my @returns = qw(idle timeout type command);
  push @returns, 'module' if $self->command() eq 'smoke';
  return map { ( $_ => $self->$_ ) } @returns;
}

1;
__END__

=head1 NAME

POE::Component::SmokeBox::Job - Object defining a SmokeBox job.

=head1 SYNOPSIS

  use POE::Component::SmokeBox::Job;

  my $job = POE::Component::SmokeBox::Job->new(
	type    => 'CPANPLUS::YACSmoke',
	command => 'smoke',
	module  => 'B/BI/BINGOS/Acme-POE-Acronym-Generator-1.14.tar.gz',
  );

=head1 DESCRIPTION

POE::Component::SmokeBox::Job is a class encapsulating L<POE::Component::SmokeBox> jobs.

=head1 CONSTRUCTOR

=over

=item C<new>

Creates a new POE::Component::SmokeBox::Job object. Takes a number of parameters:

  'idle', number of seconds before jobs are killed for idling, default 600;
  'timeout', number of seconds before jobs are killed for excess runtime, default 3600;
  'type', the type of backend to use, default is 'CPANPLUS::YACSmoke';
  'command', the command to run, 'check', 'index' or 'smoke', default is 'check';
  'module', the distribution to smoke, mandatory if command is 'smoke';
  'smokers', an arrayref of POE::Component::SmokeBox::Smoker objects;

C<smokers> is only required if you are directly using L<POE::Component::SmokeBox::JobQueue>

=back

=head1 METHODS

Accessor methods are provided via L<Object::Accessor>. 

=over

=item C<idle>

Number of seconds before jobs are killed for idling, default 600.

=item C<timeout>

Number of seconds before jobs are killed for excess runtime, default 3600

=item C<type>

The type of backend to use, default is C<'CPANPLUS::YACSmoke'>.

=item C<command>

The command to run, C<'check'>, C<'index'> or C<'smoke'>, default is C<'check'>.

=item C<module>

The distribution to smoke, mandatory if command is C<'smoke'>.

=item C<smokers>

An arrayref of L<POE::Component::SmokeBox::Smoker> objects

=item C<dump_data>

Returns all the data contained in the object as a list.

=back

=head1 AUTHOR

Chris C<BinGOs> Williams

=head1 LICENSE

Copyright C<(C)> Chris Williams

This module may be used, modified, and distributed under the same terms as Perl itself. Please see the license that came with your Perl distribution for details.

=head1 SEE ALSO 

L<POE::Component::SmokeBox>

L<POE::Component::SmokeBox::JobQueue>

L<Object::Accessor>

=cut


Makefile.PL
use strict;
use inc::Module::Install;
#my $RUNNING_IN_HELL = $^O eq 'MSWin32' ? 1 : 0;
name 'POE-Component-SmokeBox';
author 'Chris Williams <chris@bingosnet.co.uk>';
license 'perl';
perl_version '5.006';
abstract_from 'lib/POE/Component/SmokeBox.pm';
version_from 'lib/POE/Component/SmokeBox.pm';
build_requires 'Test::Harness' => 3.14;
build_requires 'Test::More' => 0.47;
build_requires 'File::Spec' => 0;
requires 'POE' => 1.0001;
requires 'POE::Wheel::Run::Win32' => 0.10 if win32; #$RUNNING_IN_HELL;
requires 'Module::Pluggable' => 0;
requires 'Params::Check' => 0;
requires 'Object::Accessor' => 0;
requires 'Digest::MD5' => 0;
requires 'Env::Sanctify' => 0;

unless ( can_use 'IO::Pty' and not win32 ) {
    requires 'IO::Pty' => 0 if
       prompt( 'Do you want to install IO::Pty, it is highly recommended that you do [Y/n]?', 'Y' )
         =~ /^Y/i;
}
else {
     warn "You already have IO::Pty installed, excellent.\n" unless win32; #$RUNNING_IN_HELL;
}

tests 't/*.t t/*/*.t';

print <<NOTICE;

\a--------------------------------------------------------------
Hi,

This is an early release of POE::Component::SmokeBox.

Thanks for your interest.

There may still be stuff wrong with it, but it should be
usable. Enjoy.

Any bug reports, feature requests, send an email to 
<bug-POE-Component-SmokeBox\@rt.cpan.org>

--------------------------------------------------------------

NOTICE

sleep(5);

WriteAll();
